crates/ide_assists/src/handlers/merge_match_arms.rs

```diff
@@ -84,10 +90,69 @@ pub(crate) fn merge_match_arms(acc: &mut Assists, ctx: &AssistContext) -> Option
     )
 }
 
-fn contains_placeholder(a: &ast::MatchArm) -> bool {
+fn contains_placeholder(a: &MatchArm) -> bool {
     matches!(a.pat(), Some(ast::Pat::WildcardPat(..)))
 }
 
+fn are_same_types(
+    current_arm_types: &HashMap<String, Option<TypeInfo>>,
+    arm: &ast::MatchArm,
+    ctx: &AssistContext,
+) -> bool {
+    let arm_types = get_arm_types(&ctx, &arm);
+    for (other_arm_type_name, other_arm_type) in arm_types {
+        if let (Some(Some(current_arm_type)), Some(other_arm_type)) =
+            (current_arm_types.get(&other_arm_type_name), other_arm_type)
+        {
+            if other_arm_type.original != current_arm_type.original {
+                return false;
+            }
+        } else {
+            // No corresponding field found
+            return false;
+        }
+    }
+
+    return true;
+}
+
+fn get_arm_types(context: &AssistContext, arm: &MatchArm) -> HashMap<String, Option<TypeInfo>> {
+    let mut mapping: HashMap<String, Option<TypeInfo>> = HashMap::new();
+
+    fn recurse(
+        pat: &Option<Pat>,
+        map: &mut HashMap<String, Option<TypeInfo>>,
+        ctx: &AssistContext,
+    ) {
+        if let Some(local_pat) = pat {
+            match pat {
+                Some(ast::Pat::TupleStructPat(tuple)) => {
+                    for field in tuple.fields() {
+                        recurse(&Some(field), map, ctx);
+                    }
+                }
+                Some(ast::Pat::RecordPat(record)) => {
+                    if let Some(field_list) = record.record_pat_field_list() {
+                        for field in field_list.fields() {
+                            recurse(&field.pat(), map, ctx);
+                        }
+                    }
+                }
+                Some(ast::Pat::IdentPat(ident_pat)) => {
+                    if let Some(name) = ident_pat.name() {
+                        let pat_type = ctx.sema.type_of_pat(local_pat);
+                        map.insert(name.text().to_string(), pat_type);
+                    }
+                }
+                _ => (),
+            }
+        }
```

Comment:
This needs to cover every pattern that can hold a child pattern so,
```
    IdentPat(IdentPat),
    BoxPat(BoxPat),
    OrPat(OrPat),
    ParenPat(ParenPat),
    RangePat(RangePat),
    RecordPat(RecordPat),
    RefPat(RefPat),
    SlicePat(SlicePat),
    TuplePat(TuplePat),
    TupleStructPat(TupleStructPat),
```

Reply:
Okay, I can look into each of these and figure something out. You mention `RangePat` -- I've already got a unit test that shows a range is supported, could you give me an example of a `RangePat` expression that isn't taken care of yet?

Similarly, if I do something like this (targeting `ParenPat`) then it also works out of the box:

```rs
match x {
	1 => "",
	((((2)))) => "",
	_ => "other"
};
```

I'll dig in deeper tonight whether I really do get it out of the box or if it's just a deficiency in some of the test scenarios I have right now. Would be keen to get your take on this though

Reply:
This is only relevant for patterns when they contain an `IdentPattern`, as in when they bind a name. So an example that would fail this particular part is something like:
```rs
match (0, "boo") {
    (x, y) => println!("{}", x),
    (y, x) => println!("{}", x),
}
```
Merging this shouldn't be allowed, as `x` and `y` each bind to different types in these arms. I don't think there is valid code for `RangePattern`s that could trigger this so there is no need to test those specifically.

Reply:
@Veykril I've updated the PR to account for the other patterns. `TuplePat`, `ParenPat`, `RefPat` and `SlicePat` are either handled or have tests showing they don't break. I couldn't come up with examples for some of the other patterns (i.e. `BoxPat` and `OrPat`) -- if you have a concrete example of these in mind then that'd be great.